src/edu/ucsb/nceas/metacat/properties/AuthPropertiesDelegate.java

```diff
@@ -0,0 +1,134 @@
+            throw new GeneralPropertyException("I/O problem while loading properties: " + ioe.getMessage());
+        }
+    }
+
+    protected static AuthPropertiesDelegate getInstance()
+        throws GeneralPropertyException {
+        if (authPropertiesDelegate == null) {
+            authPropertiesDelegate = new AuthPropertiesDelegate();
+        }
+        return authPropertiesDelegate;
+    }
+
+    /**
+     * Get the auth backup properties file. These are configurable properties that are stored
+     * outside the metacat install directories so the user does not need to re-enter all the
+     * configuration information every time they do an upgrade.
+     *
+     * @return a SortedProperties object with the backup properties
+     */
+    protected SortedProperties getAuthBackupProperties() {
+        return authBackupProperties;
+    }
+
+    /**
+     * Get the auth properties metadata. This is retrieved from an xml file that describes the
+     * attributes of configurable properties.
+     *
+     * @return a PropertiesMetaData object with the organization properties metadata
+     */
+    protected PropertiesMetaData getAuthMetaData() {
+        return authMetaData;
+    }
+
+    /**
+     * Writes out backup configurable properties to a file.
+     */
+    protected void persistAuthBackupProperties()
+        throws GeneralPropertyException {
+
+        // Use the metadata to extract configurable properties from the
+        // overall properties list, and store those properties.
+        try {
+            SortedProperties backupProperties = new SortedProperties(siteAuthPropsFilePath);
+
+            // Populate the backup properties for auth properties using
+            // the associated metadata file
+            PropertiesMetaData authMetadata = new PropertiesMetaData(authMetadataFilePath);
+
+            Map<String, MetaDataProperty> authKeyMap = authMetadata.getProperties();
+            Set<String> authKeySet = authKeyMap.keySet();
+            for (String propertyKey : authKeySet) {
+                // don't backup passwords
+                MetaDataProperty metaData = authKeyMap.get(propertyKey);
+                if (!metaData.getFieldType().equals(MetaDataProperty.PASSWORD_TYPE)) {
+                    backupProperties.addProperty(
+                        propertyKey, PropertyService.getProperty(propertyKey));
+                }
+            }
+
+            // store the properties to file
+            backupProperties.store();
+            authBackupProperties = new SortedProperties(siteAuthPropsFilePath);
+            authBackupProperties.load();
+
+        } catch (TransformerException te) {
+            throw new GeneralPropertyException(
+                "Could not transform backup properties xml: " + te.getMessage());
+        } catch (IOException ioe) {
+            throw new GeneralPropertyException(
+                "Could not backup configurable properties: " + ioe.getMessage());
+        }
```

Comment:
This looks good, but these are major problems -- do you log the failure elsewhere in the calling class?

Reply:
the Auth Admin code is the original caller. If these exceptions are thrown, they are added to the list of validation errors to be displayed to the admin user in the UI: https://github.com/NCEAS/metacat/blob/3981309e112ba1f1cd2ead6d50d61e45438cb0d6/src/edu/ucsb/nceas/metacat/admin/AuthAdmin.java#L216